scripts/nbd-firewall-config.sh

```diff
@@ -0,0 +1,91 @@
+#!/bin/bash
+#
+# Copyright (c) Citrix Systems. All rights reserved.
+
+##############################################################
+# Use this script to open/close the specified interfaces
+# (IP addresses) for incoming TCP connections to the NBD port.
+#
+# Usage:
+#   ./nbd-firewall-config set [address...]
+#
+##############################################################
+
+# Historical note: the original version of this file filtered
+# by local IP addresses rather than by network interfaces;
+# that version might one day be a useful reference for something.
+
+# Require the "set" argument because otherwise there would be a risk
+# of someone running the script with no arguments and accidentally
+# setting NBD to be available on no addresses.
+# This way, omitting "set" (or supplying "--help") causes the script
+# to print usage instructions.
+
+set -e
+OP="$1"
+if [ _"${OP}" != _set ]; then
+    echo "Usage: $(basename ${0}) set [interface...]
+will set the firewall to allow incoming TCP connections
+to the NBD port on only the specified interfaces (if any)." 1>&2
+    exit 1
+fi
+shift 1
+
+set -eu
+# No automatic cleanup on error because it would depend too
+# critically on the stage at which the error occurs.
+
+NBDPORT=10809
+# Note that chain names must be under 29 characters.
+NEWCHAININ=xapi_nbd_input_chain_new
+NBDCHAININ=xapi_nbd_input_chain
+NEWCHAINOUT=xapi_nbd_output_chain_new
+NBDCHAINOUT=xapi_nbd_output_chain
+
+# Rule fragments for sending NBD packets to NBD chains
+# from INPUT and OUTPUT chains.
+NBD_INPUT_JUMP_TO="-p tcp --dport $NBDPORT -j "
+NBD_OUTPUT_JUMP_TO="-p tcp --sport $NBDPORT -j "
+
+# Same principle as double-buffering a display:
+# add the items one by one to temporary chains,
+# then swap them into use atomically, or as near as possible.
+iptables --new-chain $NEWCHAININ 2>/dev/null || iptables --flush $NEWCHAININ
+iptables --new-chain $NEWCHAINOUT 2>/dev/null || iptables --flush $NEWCHAINOUT
+while [ "$#" -ne 0 ]; do
+    intf="${1}"
+    # (Each positional param is forgotten for ever at the end of the loop.)
+    shift 1
+    iptables -A $NEWCHAININ -m conntrack --ctstate NEW,ESTABLISHED --in-interface "${intf}" -j ACCEPT
+    iptables -A $NEWCHAINOUT --out-interface "${intf}" -j RETURN
```

Comment:
What is the significance of this rule. Why do you need an out chain ?

Reply:
Just trying to be thorough. For example, if there is an existing established NBD connection on which the server is busy sending disc data to the client, then I would like to block these outbound packets if the relevant interface is not in the newly specified set. (Yes, I know the inbound chain would block the ACK packets from the client.)

Reply:
lgtm